### PR TITLE
Update to most recent viable nightly

### LINKED ISSUE #1628: Feature request: "rustup update nightly" to the last version which has all necessary components.
Instead of failing or forcing, just use not-so-up-to-date version.

Additinoal feature request: `rustup update nightly --date=2019-01-29` to update nightly to a specific version while keeping it as "nigthly", not as a separate toolchain. Should also allow downgrades.

### PR DESCRIPTION
**Pull request by jonhoo**
Previously, if the user had components installed on `nightly` which were not available for the latest nightly, `rustup update` would not update the user's nightly at all. With this patch, rustup will try progressively older nightlies until it finds a nightly that supports all the components the user has installed for their current nightly.

Fixes #1628.
Makes progress towards #1501.
Fixes the underlying issue in #1676.

### CODE REVIEW

**Comment by jonhoo** [comment]
As an anecdotal observation, the code for mocking a `dist` in `tests/mock/clitools.rs ` should probably be refactor separately from this change. It's currently pretty messy, and would benefit from some cleanup, and probably some macros.

**Comment by jonhoo** [comment]
Cleanup in https://github.com/rust-lang/rustup.rs/pull/1998

File: tests/cli-misc.rs
Comment (najamelan) [inline tests/cli-misc.rs:11]: Just wondering about feedback. It makes sense that the old toolstate message isn't appropriate anymore, but is there a message that says something like: "Cannot update nightly channel because no newer nightlies were found that have all the required components: x, y z" and another like "Your nighty has been updated to version xxxx-xx-xx because more recent versions don't have the required components: x, y, z"?

Comment (jonhoo) [reply tests/cli-misc.rs:11]: So, you do get this info-level message:
https://github.com/rust-lang/rustup.rs/blob/e2da58a59bdf34b68fa7074d4c42f11cbc5a22f9/src/dist/notifications.rs#L164-L168
as rustup searches backwards in history, though maybe we want an additional message saying which nightly was chosen in the end? Although that said, that _should_ be clear from the final install message.

As for "Cannot update nightly channel because no newer nightlies were found that have all the required components: x, y z", the only way that could happen is if the nightly you are already on is the newest one that has all components. In that case, rustup will give the same message it currently gives if you try to update and there are no updates available. The info messages from above should make it clear why any newer nightlies weren't downloaded.

So, with that all said, do you still think it's necessary to add or change any messages?

Also, to be clear that one removal was due to a previously failing test now succeeding since an older nightly had the required component :)

Comment (najamelan) [reply tests/cli-misc.rs:11]: I didn't see that info when I looked through the diff, so I suppose that probably has enough information to avoid confusion about what's happening.  It's maybe not the most elegant way of reporting, but that's just cosmetics. It could generate quite some output if you haven't updated in say a month. So I suppose it's balancing having a polished message against the hurdle of implementing it, which I can't really estimate how much work it would be. 

If rustup prints the version actually installed, than I don't think we need an extra message for that as long as there is something that hints at versions being skipped.

To summarize, I can definitely live with this and there are tons of not-cosmetic issues that are more worthwhile putting effort in.

Thanks alot for taking this on anyway, it will be a major quality of ux upgrade in any case.

Comment (najamelan) [reply tests/cli-misc.rs:11]: actually, probably it's rare to see components fail for more than 2 weeks, so the amount of output is probably quite reasonable, even after not updating for a long time.

Comment (jonhoo) [reply tests/cli-misc.rs:11]: Yeah, I was about to say, you will only see that message as many times as there are nightlies since your component set last built successfully on nightly, no matter how old your nightly is. I'm also fairly sure that rustup prints which nightly was actually installed at the end, so we should be good on that end.

I agree with you that better cosmetics could be nice here, though I _think_ this gives us at least all the info, even if it's not presented in the cleanest way. My inclination would be to merge this and then do cosmetic fixes in follow-up work (maybe related to #1501).

**Comment by norru** [comment]
This is ace. Thank you very much!

File: src/dist/dist.rs
Comment (kinnison) [inline src/dist/dist.rs]: Strictly speaking this is iterative not recursive, but I wouldn't block merge on that nitpick alone

File: src/dist/dist.rs
Comment (kinnison) [inline src/dist/dist.rs]: My gut feeling is that this could potentially be neater done as something like the following (untested):

```rust
toolchain.date = Some(Utc.from_utc_date(
    &toolchain.date.unwrap_or_else(
        ||NaiveDate::parse_from_str(&fetched, "%Y-%m-%d").expect("Malformed manifest date"))
    )).pred().format("%Y-%m-%d").to_string());
```

Obviously better formatting would be needed, and I've not checked the borrows vs. moves in that.

File: src/dist/dist.rs
Comment (kinnison) [inline src/dist/dist.rs:624]: What if we're now requesting a dated nightly which doesn't exist?  While it's rare, sometimes we don't succeed in building a nightly and so we can't provide one for a given date.

**Review by kinnison — CHANGES_REQUESTED** [review]
Apart from my concern over missing nightlies and a possible bit of reformatting, this looks like a nice simple elegant solution to the problem.

File: src/dist/dist.rs
Comment (jonhoo) [reply src/dist/dist.rs:624]: How does that manifest in the errors? I can't seem to find an error that maps cleanly to that idea. Basically, how does the manifest change if a nightly is unavailable?

Comment (mati865) [reply src/dist/dist.rs:624]: Current nightly is unavailable so it's perfect occasion to test it :laughing:.

edit: Thanks @lzutao, GitHub completion sometimes messes up.

File: src/dist/dist.rs
Comment (jonhoo) [reply src/dist/dist.rs]: Yeah, it was split up this way from back when I hadn't added `fetched`. Simplified it even further now :)

Comment (jonhoo) [reply src/dist/dist.rs]: Yeah, it was split up this way from back when I hadn't added `fetched`. Simplified it even further now :)

File: src/dist/dist.rs
Comment (jonhoo) [reply src/dist/dist.rs:624]: Actually, this turns out to already just work, since it manifests as the `rustc` component missing (and thus hits the existing case). I've added a test.

Comment (tesuji) [reply src/dist/dist.rs:624]: Wrong person @mati865

**Review by kinnison — APPROVED** [review]
LGTM :+1:  I'll merge once I have a confirm from CI.

**Comment by vi** [comment]
When will it be available for `rustup self update`?

**Comment by kinnison** [comment]
> When will it be available for `rustup self update`?

@vi When we're ready to release 1.20.0 -- We have active tester requests out right now, and we may put further requests out on the way to 1.20.0.  Once we're good, we'll release.  I'd guess at a small (one or two) number of weeks.

**Comment by Arnavion** [comment]
@jonhoo @kinnison How would I restore the original behavior? Do I have to `rustup component remove` the components I don't mind losing before `rustup update`ing, then `rustup component add` them back? Or is there an easier way?

**Comment by najamelan** [comment]
@Arnavion one option is to check https://rust-lang-nursery.github.io/rust-toolstate/ manually and not update if not all your components are there...

**Comment by jonhoo** [comment]
@Arnavion I'm not sure I follow? The old behavior was to simply _fail_ the update. If you mean to _force_ the update, then you should be able to pass `--force`.

**Comment by Arnavion** [comment]
@jonhoo The new behavior, AFAICT from your change, is that `rustup update` will always succeed, but leave me with a nightly that is not the latest. The original behavior was that `rustup update` would fail, which my "update Rust" script would detect and handle by removing components that I don't need (clippy, rustfmt), re-running `rustup update`, then `rustup component add`ing (rustfmt) or `cargo install`ing (clippy, since it's usually fixed in git master even if the submodule in the rust repo is broken) the components back.

Yes, `--force` seems to be what I was asking for. Thanks.

**Comment by EliSnow** [comment]
@jonhoo, I was reminded of this great work today when I did a `rustup update` and saw the message `skipping nightly which is missing installed component 'rls-preview'`. Thank you!

**Pull request merged**
